RAGchain/DB/pickle_db.py

```diff
@@ -54,19 +54,24 @@ def create_or_load(self):
         else:
             self.create()
 
-    def save(self, passages: List[Passage]):
+    def save(self, passages: List[Passage], upsert: bool = False):
         """Saves the given list of Passage objects to the pickle database. It also saves the data to the Linker."""
+
         # save to pickleDB
+        if os.path.exists(self.save_path):
+            self.load()
+        if not upsert:
+            id_list = list(map(lambda x: x.id, passages))
+            duplicate_ids = [doc.id for doc in self.fetch(id_list)]
+            if len(duplicate_ids) > 0:
+                raise ValueError(f"Duplicate id {duplicate_ids} found in pickle database.")
+
         self.db.extend(passages)
```

Comment:
What happen we insert duplicate id, when upsert is `True`.
Then, in PickleDB, there will be two or more duplicate id passages.
It occurs error at `fetch` or other methods, so you must overwrite this if there are duplicate ids and upsert is `True`